added PowerShell v3 support to the module; fixed Path parameter in In…

KirkMunro opened the pull request:
- Added PowerShell v3 support to PSScriptAnalyzer.  To build for PowerShell v3, select either the "PSV3 Debug" or "PSV3 Release" configuration. PSV3 support was added using a PSV3 symbol combined with #if directives throughout the solution.  Post-build steps were modified to copy in a v3-compatible psd1 file if one of the PSV3 configurations is selected.  No core functionality was changed in the v5 version, only v3 where some Ast classes do not exist
- Fixed the Path parameter of Invoke-ScriptAnalyzer so that it properly supports PowerShell paths (i.e. Modules:\HistoryPx).  This was not working before.
- Changed the Path parameter type from a string to an array and added support for the new type.
- Added the OutputType attribute to Invoke-ScriptAnalyzer (best practice).
- Added pipeline support to the Path parameter of Invoke-ScriptAnalyzer.

msftclas commented:
Hi __@KirkMunro__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

yutingc commented:
Thank you Kirk for the contribution! These changes are all very valuable to PSScriptAnalyzer. Can you please sign the agreement before we are able to pull your request？

Thank you very much!

msftclas commented:
__@KirkMunro__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

KirkMunro commented:
Thanks!  I just finished signing the agreements, so you should be good to
go!

--
*Kirk Munro *
Poshoholic, Microsoft MVP
Blog <http://poshoholic.com/> | Twitter <http://twitter.com/poshoholic> |
LinkedIn <http://ca.linkedin.com/in/kirkmunro> | Facebook
<http://www.facebook.com/#%21/kirk.munro>

Need a PowerShell SME for a project at work? Contact me
<http://poshoholic.com/contact-me/>!

On Thu, May 7, 2015 at 9:18 PM, Yuting Chen[MSFT] <notifications@github.com>
wrote:

> Thank you Kirk for the contribution! These changes are all very valuable
> to PSScriptAnalyzer. Can you please sign the agreement before we are able
> to pull your request？
>
> Thank you very much!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PowerShell/PSScriptAnalyzer/pull/122#issuecomment-100060594>
> .
>

yutingc commented:
One question, how should we handle test cases for V3 and V5?

KirkMunro commented:
That's a great question, I had done my own testing, but I hadn't updated
your tests folder.

Looking at it right now, I would think that the tests that are v5+ could
use if ($PSVersionTable.PSVersion -ge [Version]'5.0') {...} blocks.
Thoughts?

--
*Kirk Munro *
Poshoholic, Microsoft MVP
Blog <http://poshoholic.com/> | Twitter <http://twitter.com/poshoholic> |
LinkedIn <http://ca.linkedin.com/in/kirkmunro> | Facebook
<http://www.facebook.com/#%21/kirk.munro>

Need a PowerShell SME for a project at work? Contact me
<http://poshoholic.com/contact-me/>!

On Thu, May 7, 2015 at 9:24 PM, Yuting Chen[MSFT] <notifications@github.com>
wrote:

> One question, how should we handle test cases for V3 and V5?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PowerShell/PSScriptAnalyzer/pull/122#issuecomment-100062475>
> .
>

raghushantha commented:
Thanks for the changes, Kirk. First MVP contribution!

We will take a look at this.

KirkMunro commented:
Happy to contribute.

FYI, since the earlier question about the test cases for v3 vs v5, I just
updated the tests in the Tests folder so that I can run them without error
on v3.  There were a few bugs, plus I had to prevent certain things from
running on v3/v4 that just won't work (anything that uses PowerShell
classes).  I can either submit a new, comprehensive pull request that
includes these additional changes or wait until you merge my existing pull
request and then submit a pull request for the additional changes to the
Tests folder.

--
*Kirk Munro *
Poshoholic, Microsoft MVP
Blog <http://poshoholic.com/> | Twitter <http://twitter.com/poshoholic> |
LinkedIn <http://ca.linkedin.com/in/kirkmunro> | Facebook
<http://www.facebook.com/#%21/kirk.munro>

Need a PowerShell SME for a project at work? Contact me
<http://poshoholic.com/contact-me/>!

On Thu, May 7, 2015 at 9:36 PM, Raghu Shantha [MSFT] <
notifications@github.com> wrote:

> Thanks for the changes, Kirk. First MVP contribution!
>
> We will take a look at this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PowerShell/PSScriptAnalyzer/pull/122#issuecomment-100066638>
> .
>

yutingc commented:
Hi @KirkMunro, it would be helpful if you could submit a new request that includes all the additional changes. 
Thanks!

KirkMunro commented:
I was looking last night and this morning for how to do that, and from what
I can tell, it's already done since I added my additional changes to my
GitHub fork yesterday.  When I browse to the Pull Request, it shows all
changes, including those to the Tests folder, so this looks like it is
ready to go from my side.  Please let me know if you need anything else.

--
*Kirk Munro *
Poshoholic, Microsoft MVP
Blog <http://poshoholic.com/> | Twitter <http://twitter.com/poshoholic> |
LinkedIn <http://ca.linkedin.com/in/kirkmunro> | GitHub
<http://github.com/KirkMunro> | Facebook
<http://www.facebook.com/#%21/kirk.munro>

Need a PowerShell SME for a project at work? Contact me
<http://poshoholic.com/contact-me/>!

On Fri, May 8, 2015 at 1:09 PM, Yuting Chen[MSFT] <notifications@github.com>
wrote:

> Hi @KirkMunro <https://github.com/KirkMunro>, it would be helpful if you
> could submit a new request that includes all the additional changes.
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PowerShell/PSScriptAnalyzer/pull/122#issuecomment-100301051>
> .
>

yutingc commented:
Oh right, GitHub will automatically update your pull request if your pull request is still open and changes are made in the same branch. 
We will do the validation later today and will let you know if there is anything. Thanks!

KirkMunro commented:
Hello Yuting,

I just checked in another set of changes to my fork.  This set of changes
updates the project such that you can now build one module that works for
PowerShell 3.0 or later.  I was looking at various options to enable this,
and took the easiest option I could find.  To build the module so that it
works on PowerShell 3.0, 4.0, 5.0, and later, you must do the following:

1. Build the solution using the Release configuration.
2. Build the solution using the PSV3 Release configuration.

That's it.  At that point, the PSScriptAnalyzer subfolder that is generated
will contain everything that is needed to run on PowerShell 3.0, 4.0, or
5.0.

Essentially all I did was put the 3.0/4.0-compatible DLLs into a subfolder
of the module, modified the manifest so that it loads a script module
instead of the nested binary module directly, and added the script module
file to determine which binary module to load based on the current version
of PowerShell.

While I was working on these changes, I also made a few other changes that
I had noticed needed to be done:

1. I added an output type to Get-ScriptAnalyzerRule.
2. I fixed the case of PowerShell (you guys had it written as Powershell
with a lowercase "s" throughout the solution, and that isn't the proper
case to use, so I changed it solution wide to PowerShell).
3. I cleaned up the manifest, and added specifically defined
CmdletsToExport names (best practice).
4. I added PowerShellGet metadata to the manifest (so that it can be found
using the keywords "lint", "best", and "practice", and so that the license
is easily discoverable from the PowerShell Gallery (another best practice).

Once all of these changes were complete, I ran the Pester tests both on a
Windows 10 system with PowerShell 5 and on my Windows 8.1 system with
PowerShell 4, and I received the results I was expecting from those tests
(there are still some issues in the tests, but those issues have nothing to
do with these changes and were out of scope for the work I was doing here).

From my perspective, the only thing remaining to do here is to update the
README.md file so that the build instructions properly document that you
should build two configurations: Release and PSV3 Release (or Debug and
PSV3 Debug) in order to get the appropriate DLLs built and copied into the
proper locations for the module.

I think that about wraps it up.  Please let me know your thoughts on this
update.  If you like it and want me to update the README.md file with
updated build instructions before you accept the pull request, let me know.


--
*Kirk Munro *
Poshoholic, Microsoft MVP
Blog <http://poshoholic.com/> | Twitter <http://twitter.com/poshoholic> |
LinkedIn <http://ca.linkedin.com/in/kirkmunro> | GitHub
<http://github.com/KirkMunro> | Facebook
<http://www.facebook.com/#%21/kirk.munro>

Need a PowerShell SME for a project at work? Contact me
<http://poshoholic.com/contact-me/>!

On Fri, May 8, 2015 at 1:42 PM, Yuting Chen[MSFT] <notifications@github.com>
wrote:

> Oh right, GitHub will automatically update your pull request if your pull
> request is still open and changes are made in the same branch.
> We will do the validation later today and will let you know if there is
> anything. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PowerShell/PSScriptAnalyzer/pull/122#issuecomment-100308503>
> .
>

yutingc commented:
Hi @KirkMunro, thanks for all these changes!　We will spend some time do some testing before we merge the pull request. Will let you know if we need anything else. Thanks for wrapping this issue up so quickly and we will get back to you soon.

tysonjhayes commented:
@yutingc Are we still working on getting the new branch setup? I'd love to be able to test this out on some of my scripts.

yutingc commented:
@tysonjhayes Yes it's still in our plan. We had some discussions on how we will be handling the two versions of ScriptAnalyzer before we can get this V3 version ready. We will close on this issue ASAP and will let you know when it is ready for you to test. Sorry for the delay!

raghushantha commented:
Hi Kirk.

Can you submit a new PR based on our latest development branch. We have merged the latest gallery release to this branch.

Thanks.

nobodyguy commented:
Guys,  can I help you with something? What needs to be done for PS 3 support?

raghushantha commented:
@KirkMunro  - I am updating your PR to work with v1.2.0

Still testing my changes - and fixing bugs -

https://github.com/PowerShell/PSScriptAnalyzer/tree/ScriptAnalyzerPSV3Branch

raghushantha commented:
Here is the updated PR. 

Need to figure out the Appveyor part to build for a matrix of OS (WS2012R2, WS2012R2+WMF) and Configurations (Release, PSv3 Release)
